src/dom/JsValueConversion.ts

```diff
@@ -0,0 +1,142 @@
+import {Decimal, dom, Timestamp} from "../Ion";
+import {Constructor, Value} from "./Value";
+import JSBI from "jsbi";
+import {_hasValue} from "../util";
+
+
+// When Typescript is compiling this code, dom/index.ts depends on Value.ts which in turn
+// depends on this file. This means that 'dom' is still being defined and we cannot yet
+// reference any of the Value subtypes (dom.Integer, dom.String, etc). We can sidestep this
+// issue by lazily initializing this mapping, deferring any reference to those types until
+// they are first used.
+let _domTypesByIonType: Map<IonType, Constructor> | null = null;
+
+// Lazily initializes our IonType -> Dom type constructor mapping.
+function _getDomTypesByIonTypeMap(): Map<IonType, Constructor> {
+    if (_domTypesByIonType === null) {
+        // Clob, SExpression, and Symbol are not included here as _inferType always
+        // favors Blob, List, and String respectively.
+        _domTypesByIonType = new Map<IonType,  Constructor>([
+            [IonTypes.NULL,  dom.Null],
+            [IonTypes.BOOL,  dom.Boolean],
+            [IonTypes.INT,  dom.Integer],
+            [IonTypes.FLOAT, dom.Float],
+            [IonTypes.DECIMAL, dom.Decimal],
+            [IonTypes.TIMESTAMP, dom.Timestamp],
+            [IonTypes.STRING, dom.String],
+            [IonTypes.BLOB,  dom.Blob],
+            [IonTypes.LIST, dom.List],
+            [IonTypes.STRUCT, dom.Struct],
+        ]);
+    }
+    return _domTypesByIonType;
+}
+
+// Returns a dom type constructor for the provided IonType.
+// This function is exported to assist in unit testing but is not visible at the library level.
+export function _domConstructorFor(ionType: IonType): Constructor {
+    let domConstructor = _getDomTypesByIonTypeMap().get(ionType)!;
+    if (!_hasValue(domConstructor)) {
+        throw new Error(`No dom type constructor was found for Ion type ${ionType.name}`);
+    }
+    return domConstructor;
+}
+
+// Examines the provided JS value and returns the IonType best suited to represent it.
+// This function will never return Clob, SExpression, and Symbol; it will always
+// favor Blob, List, and String respectively.
+function _inferType(value: any): IonType {
+    if (value === undefined) {
+        throw new Error("Cannot create an Ion value from `undefined`.");
+    }
+    if (value === null) {
+        return IonTypes.NULL;
+    }
+
+    let valueType: string = typeof(value);
+    switch (valueType) {
+        case "string":
+            return IonTypes.STRING;
+        case "number":
+            return Number.isInteger(value)
+                ? IonTypes.INT
+                : IonTypes.FLOAT;
+        case "boolean":
+            return IonTypes.BOOL;
+        case "object":
+            break;
+        // We `break` here to defer handling native bigints to the 'valueof JSBI' test below,
+        // which will handle native bigints correctly if Babel is being used to convert JSBI
+        // logic to native bigint logic.
```

Comment:
fwiw, my eyes expect this comment to be after `case "bigint":`, and indented at the same level as `break`

Reply:
Changed.